pydeseq2/dds.py

```diff
@@ -213,55 +226,153 @@ def __init__(
+        # of the type formulaic can recognize or a column
+        if isinstance(self.design_factors, str):
+            self.continuous_factors = continuous_factors
+            if self.design_factors not in self.obs.columns:
+                # TODO uncomment when we support all of formulaic grammar
+                # try:
+                #     Formula(self.design_factors)
+                # except FormulaInvalidError as err:
+                #     raise (
+                #         f"Design factor {self.design_factors} is not a valid formula"
+                #     ) from err
+                # Remove potential spaces around + signs
+                self.design_factors = re.sub(r"\ \+|\+\ ", "+", self.design_factors)
+                # Check that we have something like ~a + b + a:b:c:d
+                assert self.design_factors.startswith("~")
+
+                for factor in itertools.chain(
+                    *[term.split(":") for term in self.design_factors[1:].split("+")]
+                ):
+                    if factor not in self.obs.columns:
+                        raise ValueError(f"Formula contain unknown factor {factor}")
+
+                self.single_design_factors = list(
+                    {col for col in self.obs.columns if col in self.design_factors}
+                    & set(self.obs.columns)
+                )
+            else:
+                # Only one factor
+                self.design_factors = [self.design_factors]
+                self.single_design_factors = self.design_factors
+
+        elif isinstance(self.design_factors, list):
+            # If there are interacting terms in the list we'll use formulaic
+            interacting_terms = [
+                factor for factor in self.design_factors if ":" in factor
+            ]
+            if len(interacting_terms) > 0:
+                self.design_factors = "~" + "+".join(self.design_factors)
+                self.single_design_factors = list(
+                    {col for col in self.obs.columns if col in self.design_factors}
+                    & set(self.obs.columns)
+                )
+            else:
+                self.single_design_factors = self.design_factors
+        else:
+            raise ValueError("Design factors should be a string or a list of strings")
+
+        # We modify only obs object
+        if np.any(["_" in factor for factor in self.single_design_factors]):
             warnings.warn(
-                """Same factor names in the design contain underscores ('_'). They will
-                be converted to hyphens ('-').""",
+                """Same factor names in the design contain underscores ('_').
+                They will be converted to hyphens ('-').""",
                 UserWarning,
                 stacklevel=2,
             )
 
-            new_factors = replace_underscores(self.design_factors)
+            new_factors = replace_underscores(self.single_design_factors)
```

Comment:
A general comment (writing it here but applies elsewhere): are there cases where we apply `replace_underscores` and then use `formulaic`? If so, we could run into trouble because (if I'm not mistaken) `-` would be interpreted as a minus sign (https://matthewwardrop.github.io/formulaic/guides/grammar/).

Reply:
It's solved I think I added logic to handle that